napari/utils/transforms/transform_utils.py

```diff
@@ -29,51 +29,220 @@ def compose_linear_matrix(rotate, scale, shear) -> np.array:
+
+def infer_ndim(scale=None, translate=None, rotate=None, shear=None):
+    """Infer the dimensionality of a transformation from its input components.
+
+    This is most useful when the dimensions of the inputs do not match, either
+    when broadcasting is desired or when an input represents a parameterization
+    of a transform component (e.g. rotate as an angle of a 2D rotation).
+
+    Parameters
+    ----------
+    rotate : float, 3-tuple of float, or n-D array.
+        If a float convert into a 2D rotation matrix using that value as an
+        angle. If 3-tuple convert into a 3D rotation matrix, using a yaw,
+        pitch, roll convention. Otherwise assume an nD rotation. Angles are
+        assumed to be in degrees. They can be converted from radians with
+        np.degrees if needed.
+    translate : 1-D array
+        A 1-D array of factors to shift each axis by. Translation is broadcast
+        to 0 in leading dimensions, so that, for example, a translation of
+        [4, 18, 34] in 3D can be used as a translation of [0, 4, 18, 34] in 4D
+        without modification. An empty translation vector implies no
+        translation.
+    scale : 1-D array
+        A 1-D array of factors to scale each axis by. Scale is broadcast to 1
+        in leading dimensions, so that, for example, a scale of [4, 18, 34] in
+        3D can be used as a scale of [1, 4, 18, 34] in 4D without modification.
+        An empty translation vector implies no scaling.
+    shear : 1-D array or n-D array
+        Either a vector of upper triangular values, or an nD shear matrix with
+        ones along the main diagonal.
+
+    Returns
+    -------
+    ndim : int
+        The maximum dimensionality of the component inputs.
+    """
+    ndim = 0
+    if scale is not None:
+        ndim = max(ndim, len(scale))
+    if translate is not None:
+        ndim = max(ndim, len(translate))
+    if rotate is not None:
+        ndim = max(ndim, _make_rotate_mat(rotate).shape[0])
+    if shear is not None:
+        ndim = max(ndim, _make_shear_mat(shear).shape[0])
+    return ndim
+
+
+def coerce_translate(translate, ndim):
+    """Coerce a translate input into an n-dimensional transform component.
+
+    Parameters
+    ----------
+    translate : 1-D array
+        A 1-D array of factors to shift each axis by. Translation is broadcast
+        to 0 in leading dimensions, so that, for example, a translation of
+        [4, 18, 34] in 3D can be used as a translation of [0, 4, 18, 34] in 4D
+        without modification. An empty translation vector implies no
+        translation.
+    ndim : int
+        The desired dimensionality of the output transform component.
+
+    Returns
+    -------
+    np.ndarray
+        The translate component as a 1D numpy array of length ndim.
+    """
+    translate_arr = np.zeros(ndim)
+    if translate is not None:
+        translate_arr[-len(translate) :] = translate
+    return translate_arr
+
+
+def coerce_scale(scale, ndim):
```

Comment:
Same as above — consider `broadcast` instead of `coerce`, and making `ndim` a keyword-only argument.